File: 0229-majority-element-ii/0229-majority-element-ii.cpp
class Solution {
public:
    vector<int> majorityElement(vector<int>& nums) {
        int p1 = nums[0];
        int p2 = nums[0];
        int c1=1, c2=0;
        int n = nums.size();
        
        if(n < 2) return nums;
        
        vector<int> ans;
        
        for(int i=1; i<nums.size(); i++){
            if(p1 == nums[i]) c1++;
            else if(p2 == nums[i])c2++;
            else{
                if(c1 == 0){
                    p1 = nums[i];
                    c1++;
                }else if(c2 == 0){
                    p2 = nums[i];
                    c2++;
                }else{
                    c1--;
                    c2--;
                }
            }
        }
        
        int a = 0, b = 0;
        
        for(int i=0; i<nums.size(); i++){
            if(p1 == nums[i]) a++;
            else if(p2 == nums[i])b++;
        }
        
        if(a > (n/3)) ans.push_back(p1);
        
        if(b > (n/3)) ans.push_back(p2);
        
        
        return ans;
    }
};

File: 0229-majority-element-ii/0229-majority-element-ii.py
class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        if len(nums)==1:
            return nums
        
        e1 = float('inf')
        e2 = float('inf')
        c1 = 0
        c2 = 0
        
        for i in range(0, len(nums)):
            if c1==0 and nums[i]!=e2:
                e1=nums[i]
                c1 += 1
            elif c2==0 and nums[i]!=e1:
                e2=nums[i]
                c2 += 1
            elif nums[i]==e1:
                c1 += 1
            elif nums[i]==e2:
                c2 += 1
            else:
                c1 -= 1
                c2 -= 1
        
        
        ans=[]
        c1, c2=0, 0
        
        for i in range(0, len(nums)):
            if nums[i]==e1:
                c1+=1
            if nums[i]==e2:
                c2+=1
        
        check = (len(nums)//3) + 1
        
        print(check)
        
        if c1>=check:
            ans.append(e1)
        
        if c2>=check:
            ans.append(e2)
            
        ans.sort()
        
        return ans
                